golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go

```diff
@@ -0,0 +1,696 @@
+		exportRetrieved:     false,
+		exportDone:          make(chan error, 1),
+	}
+	activeOperation = operationDetails
+
+	go func() {
+		var err error
+		defer func() {
+			if err != nil {
+				operationDetails.exportDone <- err
+			}
+			close(operationDetails.exportDone)
+		}()
+
+		action := &swingStoreExportAction{
+			Type:        "SWING_STORE_EXPORT",
+			BlockHeight: blockHeight,
+			Request:     "initiate",
+			Args:        []json.RawMessage{encodedExportOptions},
+		}
+
+		// blockingSend for SWING_STORE_EXPORT action is safe to call from a goroutine
+		_, err = exportsHandler.blockingSend(action, false)
+
+		if err != nil {
+			// First indicate an export is no longer in progress if the call to
+			// WaitUntilSwingStoreExportStarted has't happened yet.
+			// Then signal the current export operation if a call to
+			// WaitUntilSwingStoreExportStarted was already waiting.
+			operationDetails.exportStartedResult <- err
+			close(operationDetails.exportStartedResult)
+			logger.Error("failed to initiate swing-store export", "err", err)
+			return
+		}
+
+		// Signal that the export operation has started in the goroutine. Calls to
+		// WaitUntilSwingStoreExportStarted will no longer block.
+		close(operationDetails.exportStartedResult)
+
+		// The user provided ExportStarted function should call retrieveExport()
+		var retrieveErr error
+		err = eventHandler.ExportInitiated(blockHeight, func() error {
+			activeOperationDetails := activeOperation
+			if activeOperationDetails != operationDetails || operationDetails.exportRetrieved {
+				// shouldn't happen, but return an error if it does
+				return errors.New("export operation no longer active")
+			}
+
+			retrieveErr = exportsHandler.retrieveExport(eventHandler.ExportRetrieved)
+
+			return retrieveErr
+		})
+
+		// Restore any retrieve error swallowed by ExportStarted
+		if err == nil {
+			err = retrieveErr
+		}
+		if err != nil {
+			logger.Error("failed to process swing-store export", "err", err)
+		}
+
+		// Check whether the JS generated export was retrieved by eventHandler
+		if operationDetails.exportRetrieved {
+			return
+		}
+
+		// Discarding the export so invalidate retrieveExport
+		operationDetails.exportRetrieved = true
+
+		action = &swingStoreExportAction{
+			Type:        "SWING_STORE_EXPORT",
+			BlockHeight: blockHeight,
+			Request:     "discard",
+		}
+		_, discardErr := exportsHandler.blockingSend(action, false)
+
+		if discardErr != nil {
+			logger.Error("failed to discard swing-store export", "err", err)
+		}
+
+		if err == nil {
+			err = discardErr
+		} else if discardErr != nil {
+			// Safe to wrap error and use detailed error info since this error
+			// will not go back into swingset layers
+			err = sdkerrors.Wrapf(err, "failed to discard swing-store export after failing to process export: %+v", discardErr)
+		}
+	}()
+
+	return nil
+}
+
+// retrieveExport retrieves an initiated export and calls exportRetrieved with
+// the retrieved export.
```

Comment:
```suggestion
// retrieveExport retrieves an initiated export by sending a bridge message and on success
// calls exportRetrieved with the retrieved export.
//
// The bridge message returns a string giving the name of a directory containing the export
// files in a format described in <please insert reference here>.
//
// After calling exportRetrieved, the directory and its contents are deleted.
```

...and see below about renaming `exportRetrieved`.